### DISCUSSION TITLE
Panel Completions

### DISCUSSION THREAD

**Discussion by kassick** [emacs-lsp/lsp-mode#4661] [Ideas]
Copilot support a `textDocument/copilotPanelCompletion` call that generates a number of code suggestions at point. Unlike inline completions, these completions are supposed to be displayed in a separate window, so the user can inspect the suggestions more thoroughly.

Also unlike inline completions, `textDocument/copilotPanelCompletion` is a client-to-server notification -- there's no response. The code suggestions are provided via `$/progress` server-to-client notifications. It's up to us to collect these suggestions to then display them.

I've implemented this in `lsp-copilot` and the results can be seen below:

https://github.com/user-attachments/assets/b0af499b-9a53-4a58-9c66-45294ceaa290

- I've fixed the progress callback of copilot-ls to actually integrate with the LSP native progress reports (notice the progress in the status line)
- Completions are shown in a read-only org buffer, so the user can interact with it, copy only the parts they find relevant, etc.
- Navigation with `C-n`/`C-p` -- next / previous suggestion
- Emacs buttons available for accepting or copying the suggestion. Tab jumps to the next button
- `C-<return>` on any part of a suggestion accepts it
- Some buffer-state handling to avoid inserting stuff on modified buffers -- the server takes its sweet time providing the responses.

I'm not too sure if this functionality should live inside `lsp-mode`'s implementation of copilot itself or if it should be provided as some extra package, being so copilot-specific (though other clients implement some server-specific features, so that would not be too way off).

As I see, we can either:
- Bring this functionality to `lsp-mode`
- Update copilot-ls's progress function to call a hook -- so we can implement the panel functionality in a separate, copilot-specific package

What do you folks think? Specially @kiennq , who has so kindly reviewed all my copilot PRs so far :)

---

BTW big kudos for the folks of https://github.com/copilot-emacs/copilot.el who implemented this functionality before -- it saved me a lot of time understanding how to communicate with the server and what the payloads should be like  :heart:

**Comment by kassick** [comment]
Also: We could use this same panel code to have a "Pop to Panel" functionality in inline completions -- so the user can further study the suggestion before accepting or rejecting it.

**Reply by kassick** [reply]
I currently have something like this in my config:
```elisp
(defun lsp-inline-completion--save-tick ()
  (setq-local lsp-inline-completion--tick (buffer-chars-modified-tick)))

(defun lsp-inline-completion-pop-to-panel ()
  (interactive)
  (lsp-inline-completion--clear-overlay)
  (lsp-copilot--panel-display-buffer (buffer-name) lsp-inline-completion--items lsp-inline-completion--tick))

(add-hook 'lsp-before-inline-completion-hook #'lsp-inline-completion--save-tick)

(define-key lsp-inline-completion-active-map
                (kbd "M-*")
                #'lsp-inline-completion-pop-to-panel)
```

**Comment by kassick** [comment]
Code (WIP) for this functionality: https://github.com/emacs-lsp/lsp-mode/pull/4678

**Reply by ivucica** [reply]
Did this make it in given #4678 is closed?

And would it make sense that this uses `lsp-treemacs` in some way?

**Reply by kassick** [reply]
It didn't make into master -- I'm currently using a fork [here](https://github.com/kassick/lsp-mode/tree/fix/inline-comlpetion-keymap-v2) if you want to test it; should open PRs to upstream it soon.

How would treemacs be used?

**Reply by ivucica** [reply]
I'm not certain about full capabilities of Treemacs, but given that it seems to be used for things like diagnostics (seemingly, a selectable outline view / listview) or display of symbols, it seemed like it might make sense to choose one of the offered alternatives.

I'm not certain if it's a good idea. Honestly, now seeing the screenshot from lsp-mode's own readme (featuring diagnostics done via flycheck + lsp-ui, it seems), I'm not sure lsp-treemacs would make sense. The thought mainly came out of using the lsp-inlinecompletions (thanks for working on that!) and wondering "why can't I see multiple alternatives and have to use additional keybindings to navigate through the alternatives?".

I'll give the fork a shot when I manage to find some time -- thanks for working on this.

**Reply by kassick** [reply]
new PR opened https://github.com/emacs-lsp/lsp-mode/pull/4844 with the copilot panel feature
